lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -1311,40 +1318,79 @@ InferResetsPass::collectAnnos(FModuleOp module) {
 
   // Consume any reset annotations on module ports.
   Value reset;
-  AnnotationSet::removePortAnnotations(module, [&](unsigned argNum,
-                                                   Annotation anno) {
-    Value arg = module.getArgument(argNum);
-    if (anno.isClass(fullAsyncResetAnnoClass)) {
-      if (!isa<AsyncResetType>(arg.getType())) {
-        mlir::emitError(arg.getLoc(), "'FullAsyncResetAnnotation' must "
-                                      "target async reset, but targets ")
+  // Helper for checking annotations and determining the reset
+  auto checkAnnotations = [&](Annotation anno, Value arg) {
+    if (anno.isClass(fullResetAnnoClass)) {
+      ResetKind expectedResetKind;
+      if (auto rt = anno.getMember<StringAttr>("resetType")) {
+        if (rt == "sync") {
+          expectedResetKind = ResetKind::Sync;
+        } else if (rt == "async") {
+          expectedResetKind = ResetKind::Async;
+        } else {
+          mlir::emitError(arg.getLoc(),
+                          "'FullResetAnnotation' requires resetType == 'sync' "
+                          "| 'async', but got resetType == ")
+              << rt;
+          anyFailed = true;
+          return true;
+        }
+      } else {
+        mlir::emitError(arg.getLoc(),
+                        "'FullResetAnnotation' requires resetType == "
+                        "'sync' | 'async', but got no resetType");
+        anyFailed = true;
+        return true;
+      }
+      auto badType = false;
+      if (expectedResetKind == ResetKind::Async &&
+          !isa<AsyncResetType>(arg.getType())) {
+        badType = true;
+      } else if (expectedResetKind == ResetKind::Sync) {
+        if (!isa<UIntType>(arg.getType()) ||
+            cast<UIntType>(arg.getType()).getWidth() != 1) {
+          badType = true;
+        }
+      }
+      if (badType) {
+        auto kind = resetKindToStringRef(expectedResetKind);
+        mlir::emitError(arg.getLoc(),
+                        "'FullResetAnnotation' with resetType == '")
+            << kind << "' must target " << kind << " reset, but targets "
```

Comment:
You can probably simplify this a bunch:
```suggestion
      // Check that the type is well-formed
      bool isAsync = expectedResetKind == ResetKind::Async;
      bool validUint = false;
      if (auto uintT = dyn_cast<UIntType>(arg.getType()))
        validUint = uintT.getWidth() == 1;

      if ((isAsync && !isa<AsyncResetType>(arg.getType())) ||
          (!isAsync && !validUint)) {
        auto kind = resetKindToStringRef(expectedResetKind);
        mlir::emitError(arg.getLoc(),
                        "'FullResetAnnotation' with resetType == '")
            << kind << "' must target " << kind << " reset, but targets "
```

Reply:
Thanks!